vscode/oso/client/src/telemetry.ts

```diff
@@ -0,0 +1,135 @@
+
+type MixpanelLoadEvent = {
+  event: 'load';
+  properties: {
+    diagnostics: number;
+    errors: number;
+    has_resource_blocks: boolean;
+    successful: boolean;
+    warnings: number;
+  };
+};
+
+type MixpanelDiagnosticEvent = {
+  event: 'diagnostic';
+  properties: {
+    code: Diagnostic['code'];
+  };
+};
+
+type MixpanelMetadata = {
+  // One-way hash of VSCode machine ID.
+  distinct_id: string;
+  // Unique ID for a `diagnostic_load` call. We use this to tie diagnostic
+  // events (errors & warnings) to the load event they came from.
+  load_id: string;
+  // One-way hash of workspace folder URI.
+  workspace_id: string;
+};
+
+type MixpanelEvent = { properties: MixpanelMetadata } & (
+  | MixpanelLoadEvent
+  | MixpanelDiagnosticEvent
+);
+
+type State = ExtensionContext['globalState'];
+
+export function flushQueue(
+  state: State,
+  outputChannel: OutputChannel
+): () => void {
+  return () => {
+    void (async () => {
+      try {
+        // Retrieve all queued events.
+        const events = state.get<MixpanelEvent[]>(telemetryEventsKey, []);
+
+        if (events.length === 0) return;
+
+        // Clear events queue.
+        outputChannel.appendLine(`Flushing ${events.length.toString()} events`);
+        await state.update(telemetryEventsKey, []);
+
+        mixpanel.track_batch(events, errors =>
+          errors.forEach(({ name, message, stack }) => {
+            outputChannel.appendLine(`Mixpanel track_batch error: ${name}`);
+            outputChannel.appendLine(`\t${message}`);
+            if (stack) outputChannel.appendLine(`\t${stack}`);
+          })
+        );
+      } catch (e) {
+        outputChannel.append('Caught error while sending telemetry: ');
+        outputChannel.appendLine(e);
+      }
+    })();
+  };
+}
+
+export type TelemetryEvent = {
+  diagnostics: Diagnostic[];
+  has_resource_blocks: boolean;
+};
+
+export function enqueueEvent(
+  state: State,
+  log: OutputChannel,
+  uri: Uri,
+  { diagnostics, has_resource_blocks }: TelemetryEvent
+): void {
+  void (async () => {
+    try {
+      const load_id = hash(Math.random().toString());
+      const workspace_id = hash(uri.toString());
+      const metadata: MixpanelMetadata = { distinct_id, load_id, workspace_id };
+
+      const errors = diagnostics.filter(d => d.severity === Severity.Error);
+      const warnings = diagnostics.filter(d => d.severity === Severity.Warning);
+
+      const loadEvent: MixpanelEvent = {
+        event: 'load',
+        properties: {
+          diagnostics: diagnostics.length,
+          errors: errors.length,
+          has_resource_blocks,
+          successful: errors.length === 0,
+          warnings: warnings.length,
+          ...metadata,
+        },
+      };
+
+      const diagnosticEvents: MixpanelEvent[] = diagnostics.map(({ code }) => ({
+        event: 'diagnostic',
+        properties: { code, ...metadata },
+      }));
+
+      // TODO(gj): race condition?
```

Comment:
Yes, right? The get & update are not atomic so another async task could come in & do a complete update which would not be included in our `old` list and would then be overwritten.

Reply:
Yeah, I didn't really care when I wrote this since it isn't the end of the world if we don't capture every single piece of telemetry. (Relevant: many telemetry implementations out there seem to set a "filter" that will only send a sample of X% of collected telemetry.) Especially now that #1414 sends telemetry on VSCode shutdown instead of on startup, there's no reason to use the Memento API at all. I'd prefer to ship this as-is and rip it out in #1414.

Reply:
Removed in #1414